nomad/leader.go

```diff
@@ -167,6 +176,44 @@ func (s *Server) restoreEvalBroker() error {
 	return nil
 }
 
+// restorePeriodicDispatcher is used to restore all periodic jobs into the
+// periodic dispatcher. It also determines if a periodic job should have been
+// created during the leadership transition and force runs them. The periodic
+// dispatcher is maintained only by the leader, so it must be restored anytime a
+// leadership transition takes place.
+func (s *Server) restorePeriodicDispatcher() error {
+	iter, err := s.fsm.State().JobsByPeriodic(true)
+	if err != nil {
+		return fmt.Errorf("failed to get periodic jobs: %v", err)
+	}
+
+	now := time.Now()
+	for i := iter.Next(); i != nil; i = iter.Next() {
+		job := i.(*structs.Job)
+		s.periodicDispatcher.Add(job)
+
+		launch, err := s.fsm.State().PeriodicLaunchByID(job.ID)
+		if err != nil || launch == nil {
+			return fmt.Errorf("failed to get periodic launch time: %v", err)
+		}
+
+		if !job.Periodic.Next(launch.Launch).Before(now) {
```

Comment:
Can you add some more comments here? This is a really important line but its pretty hard to reason about it especially without any supporting comments